### PR TITLE
SOLR-12217: Support shards.preference for individual shard requests

### PR DESCRIPTION
**Pull request by HoustonPutman**
# Description

Use shard preference rules when choosing replicas to send individual shard requests. This includes requests from SolrJ as well as requests sent via streaming expressions.

# Solution
How routing preferences are determined:

For Streaming Expressions, it will use the preferences from the following sources, in order of preference (defaults last):
1. The params given in the stream source
1. The URL params sent in the streaming request
1. The `DEFAULT_SHARD_PREFERENCES` set in cluster properties

For SolrJ, it will use the preferences from the params of the request it is sending.

# Tests

Current tests pass.

**TODO** Add tests for routing functionality

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I am authorized to contribute this code to the ASF and have removed any code I do not have a license to distribute.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `ant precommit` and the appropriate test suite.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: solr/solrj/src/java/org/apache/solr/client/solrj/impl/BaseCloudSolrClient.java [hunk 1]
```diff
@@ -1109,11 +1125,12 @@ public RouteException(ErrorCode errorCode, NamedList<Throwable> throwables, Map<
         }
       }
 
-      // Shuffle the leaders, if any    (none if !sendToLeaders)
+      // Sort the leader replicas, if any, according to the request preferences    (none if !sendToLeaders)
+      replicaListTransformer.transform(replicas);
```
Comment (tflobbe) [inline solr/solrj/src/java/org/apache/solr/client/solrj/impl/BaseCloudSolrClient.java]: `theUrlList`?

File: solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/SignificantTermsStream.java [hunk 2]
```diff
@@ -282,6 +281,7 @@ public Explanation toExplanation(StreamFactory factory) throws IOException {
   }
 
   public Tuple read() throws IOException {
+
```
Comment (tflobbe) [inline solr/solrj/src/java/org/apache/solr/client/solrj/io/stream/SignificantTermsStream.java]: Maybe remove this file from the PR

**Comment by HoustonPutman** [comment]
Want to hear your opinion @tflobbe .

So I changed the logic for [non-update/admin/v2 request routing](https://github.com/apache/lucene-solr/pull/984/files#diff-e188c2b5189b7a912c7c77b50beddf69L1102) to use core URLs instead of collection URLs, unless the request is querying multiple collections. In this way, if there are certain replicas on a host that you want to target you can make that distinction. However that does change the default behavior in a non-trivial way.

The only place where I see this breaks any test is weirdly enough in the [Hadoop auth tests for sending a commit request](https://github.com/apache/lucene-solr/blob/master/solr/core/src/test/org/apache/solr/security/hadoop/TestSolrCloudWithHadoopAuthPlugin.java#L127). It seems that since the commit request is sent directly to a core, instead of the collection url, it is missing an additional count of unauthenticated requests.

**Comment by HoustonPutman** [comment]
I've decided to be conservative and revert that logic. So for non-update requests, SolrJ will send the request to the collection URL, not the replica URL. All tests pass on my end now.

**Pull request merged**